File: 프로그래머스/lv3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    int n;
    int[] times;
    long answer, mid;
    public long solution(int n, int[] times) {
        this.n = n;
        this.times = times;
        Arrays.sort(times);
        binarySearch(0, Long.MAX_VALUE / 10);
        return answer;
    }
    long cnt; // mid 시간동안 받을 수 있는 최대 인원 수
    void binarySearch(long start, long end) {
        while (start <= end) {
            cnt = 0;
            mid = start + (end-start)/2;
            
            for (int time: times) {
                cnt += mid / time;
            }
            
            if (cnt < n) {
                start = mid + 1;
            } else {
                answer = mid;
                end = mid - 1;
            }
        }
        
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    left, right = 1, min(times) * n
    while left <= right:
        mid = (left + right) // 2
        count = 0
        for time in times:
            count += mid // time
        if count < n:
            left = mid + 1
        else:
            right = mid - 1
            answer = mid
            
    return answer